AEPNotificationContent/Sources/AEPNotificationViewController.swift

```diff
@@ -14,48 +14,93 @@ import UIKit
 import UserNotifications
 import UserNotificationsUI
 
-open class AEPNotificationViewController: UIViewController, UNNotificationContentExtension {
+open class AEPNotificationViewController: UIViewController, UNNotificationContentExtension, TemplateControllerDelegate {
+    var notification: UNNotification?
+
+    // MARK: - UNNotificationContentExtension delegate methods
+
     /// This method is called in the main thread of the notification content app extension
     public func didReceive(_ notification: UNNotification) {
-        guard let payload = notification.request.content.userInfo as? [String: AnyObject],
-              let template = payload[AEPNotificationContentConstants.PayloadKey.TEMPLATE_TYPE] as? String
+        self.notification = notification
+        let notificationContent = notification.request.content
+
+        // Extract the template type from the notification content
+        guard let notificationUserInfo = notification.request.content.userInfo as? [String: AnyObject],
+              let templateType = notificationUserInfo[AEPNotificationContentConstants.PayloadKey.TEMPLATE_TYPE] as? String,
+              let payload = createPayload(for: templateType, from: notificationContent)
         else {
-            displayFallbackTemplate(notification)
+            displayFallbackTemplate(notificationContent)
             return
         }
 
-        switch template {
-        case AEPNotificationContentConstants.PayloadKey.TemplateType.BASIC:
-            if let payload = BasicPayload(from: notification.request.content) {
-                let basicTemplateController = BasicTemplateController(payload)
-                view.addSubview(basicTemplateController.view)
-                preferredContentSize.height = basicTemplateController.preferredContentSize.height
-            } else {
-                displayFallbackTemplate(notification)
-            }
-        case AEPNotificationContentConstants.PayloadKey.TemplateType.CAROUSEL:
-            print("//TODO: Create and display Carousel viewController")
-        case AEPNotificationContentConstants.PayloadKey.TemplateType.TIMER:
-            print("//TODO: Create and display Timer viewController")
-        default:
-            displayFallbackTemplate(notification)
+        // Create a controller based on the template type
+        let controller: UIViewController?
```

Comment:
initialize to `nil` then you can skip the `else` clause below

Reply:
The reason why I dont assign it to nil is because I want to have the controller as a let constant. I want them to be assigned only once and don't give chance for the code to re-assign controller value again


I have refactored the code little bit.. Let me know how it looks
```
       /// Create a controller based on the template type
        let controller: UIViewController
        if let basicPayload = payload as? BasicPayload {
            controller = BasicTemplateController(withPayload: basicPayload, delegate: self)
        } else {
            controller = FallbackTemplateController(notificationContent: notificationContent)
        }

        /// Add the controller and its view as a child to the AEPNotificationViewController
        addChild(controller)
        view.addSubview(controller.view)        
    }

```